transaction/builder/src/memo_builder/rth_memo_builder.rs

```diff
@@ -82,17 +82,41 @@ pub enum CustomMemoType {
     FlexibleMemoGenerator(FlexibleMemoGenerator),
 }
 
-/// A function that returns a flexible memopayload for outputs
-pub type FlexibleOutputMemoGenerator = Box<
-    dyn Fn(FlexibleMemoOutputContext) -> Result<FlexibleMemoPayload, NewMemoError> + Sync + Send,
->;
-/// A function that returns a flexible memopayload for change
-pub type FlexibleChangeMemoGenerator = Box<
-    dyn Fn(FlexibleMemoChangeContext) -> Result<FlexibleMemoPayload, NewMemoError> + Sync + Send,
->;
+/// This is a trait that must be implemented by generators used for creating
+/// flexible output memos and flexible change memos
+pub trait FlexibleMemoGeneratorTrait: Sync + Send + Debug {
+    /// This is called when generating output memos.
+    /// It must return a memo payload with the first byte being 0x01
+    fn create_output_memo(
+        &self,
+        context: FlexibleMemoOutputContext,
+    ) -> Result<FlexibleMemoPayload, NewMemoError>;
+    /// This is called when generating change memos.
+    /// It must return a memo payload with the first byte being 0x02
+    fn create_change_memo(
+        &self,
+        context: FlexibleMemoChangeContext,
+    ) -> Result<FlexibleMemoPayload, NewMemoError>;
+}
+
+/// This is a struct that contains a generator used for creating flexible output
+/// memos and flexible change memos
+#[derive(Clone)]
+pub struct FlexibleMemoGenerator {
+    /// This is a reference to the actual generator that will be called.
+    pub generator: Arc<Box<dyn FlexibleMemoGeneratorTrait>>,
+}
```

Comment:
Traits are meant to be implemented by structs via impl blocks. This `FlexibleMemoGenerator` struct adds a layer of indirection- it holds a dynamically erased type that implements the trait, which is not commonly done in rust.

Remove this struct and see below for how to remove this indirection.

Reply:
I'm not sure I understand how to do this. Do you mind helping me out? I ended up using this approach because it seemed like the way forward to get clone on the containing struct which I want.

Reply:
I agree, we shouldn't introduce this struct, this is needlessly complicated

Reply:
I'm working on it. Should be updated shortly.